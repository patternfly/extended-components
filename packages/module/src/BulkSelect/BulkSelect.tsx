import React, { useMemo, useState } from 'react';
import {
  Dropdown,
  DropdownItem,
  DropdownList,
  DropdownProps,
  MenuToggle,
  MenuToggleCheckbox,
  MenuToggleCheckboxProps,
  MenuToggleElement
} from '@patternfly/react-core';

export const BulkSelectValue = {
  all: 'all',
  none: 'none',
  page: 'page',
  nonePage: 'nonePage'
} as const;

export type BulkSelectValue = typeof BulkSelectValue[keyof typeof BulkSelectValue];

export interface BulkSelectProps extends Omit<DropdownProps, 'toggle' | 'onSelect'> {
  /** BulkSelect className */
  className?: string;
  /** Indicates whether selectable items are paginated */
  isDataPaginated?: boolean;
  /** Indicates whether "Select all" option should be available */
  canSelectAll?: boolean;
  /** Number of entries present in current page */
  pageCount?: number;
  /** Number of selected entries */
  selectedCount: number;
  /** Number of all entries */
  totalCount?: number;
  /** Indicates if ALL current page items are selected */
  pageSelected?: boolean;
  /** Indicates if ONLY some current page items are selected */
  pagePartiallySelected?: boolean;
  /** Callback called on item select */
  onSelect: (value: BulkSelectValue) => void;
  /** Custom OUIA ID */
  ouiaId?: string;
  /** Additional props for MenuToggleCheckbox */
  menuToggleCheckboxProps?: Omit<MenuToggleCheckboxProps, 'onChange' | 'isChecked' | 'instance' | 'ref'>;
}

export const BulkSelect: React.FC<BulkSelectProps> = ({
  isDataPaginated = true,
  canSelectAll,
  pageSelected,
  pagePartiallySelected,
  pageCount,
  selectedCount = 0,
  totalCount,
  ouiaId = 'BulkSelect',
  onSelect,
  menuToggleCheckboxProps,
  ...props
}: BulkSelectProps) => {
  const [ isOpen, setOpen ] = useState(false);

  const splitButtonDropdownItems = useMemo(
    () => (
      <>
        <DropdownItem ouiaId={`${ouiaId}-select-none`} value={BulkSelectValue.none} key={BulkSelectValue.none}>
          Select none (0)
        </DropdownItem>
        {isDataPaginated && (
          <DropdownItem ouiaId={`${ouiaId}-select-page`} value={BulkSelectValue.page} key={BulkSelectValue.page}>
            {`Select page${pageCount ? ` (${pageCount})` : ''}`}
          </DropdownItem>
        )}
        {canSelectAll && (
          <DropdownItem ouiaId={`${ouiaId}-select-all`} value={BulkSelectValue.all} key={BulkSelectValue.all}>
            {`Select all${totalCount ? ` (${totalCount})` : ''}`}
          </DropdownItem>
        )}
      </>
    ),
    [ isDataPaginated, canSelectAll, ouiaId, pageCount, totalCount ]
  );

  const allOption = isDataPaginated ? BulkSelectValue.page : BulkSelectValue.all;
  const noneOption = isDataPaginated ? BulkSelectValue.nonePage : BulkSelectValue.none;

  return (
    <Dropdown
      shouldFocusToggleOnSelect
      ouiaId={`${ouiaId}-dropdown`}
      onSelect={(_e, value) => {
        setOpen(!isOpen);
        onSelect?.(value as BulkSelectValue);
      }}
      isOpen={isOpen}
      onOpenChange={(isOpen: boolean) => setOpen(isOpen)}
      toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
        <MenuToggle
          ref={toggleRef}
          isExpanded={isOpen}
          onClick={() => setOpen(!isOpen)}
          aria-label="Bulk select toggle"
          data-ouia-component-id={`${ouiaId}-toggle`}
          splitButtonOptions={{
            items: [
              <MenuToggleCheckbox
                ouiaId={`${ouiaId}-checkbox`}
                id={`${ouiaId}-checkbox`}
                key="bulk-select-checkbox"
                aria-label={`Select ${allOption}`}
                isChecked={
                  (isDataPaginated && pagePartiallySelected) ||
                  (!isDataPaginated && selectedCount > 0)
                    ? null
                    : pageSelected || selectedCount === totalCount
                }
                onChange={(checked) => onSelect?.(!checked || checked === null ? noneOption : allOption)}
                {...menuToggleCheckboxProps}
              />,
              selectedCount > 0 ? (
                <span data-ouia-component-id={`${ouiaId}-text`} key="bulk-select-text">
                  {`${selectedCount} selected`}
                </span>
              ) : null
            ]
          }}
        />
      )}
      {...props}
    >
      <DropdownList>{splitButtonDropdownItems}</DropdownList>
    </Dropdown>
  );
};

export default BulkSelect;
