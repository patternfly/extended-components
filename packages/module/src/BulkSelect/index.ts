export { default } from './BulkSelect';
export * from './BulkSelect';
