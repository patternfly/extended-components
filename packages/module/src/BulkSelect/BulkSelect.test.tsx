import React from 'react';
import { render } from '@testing-library/react';
import BulkSelect from './BulkSelect';

describe('BulkSelect component', () => {
  test('should render', () => {
    expect(render(
      <BulkSelect
        canSelectAll
        pageCount={5}
        totalCount={10}
        selectedCount={2}
        pageSelected={false}
        pagePartiallySelected={true}
        onSelect={() => null}
      />)).toMatchSnapshot();
  });
});