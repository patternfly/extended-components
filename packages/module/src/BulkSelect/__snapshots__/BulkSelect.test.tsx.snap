// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkSelect component should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pf-v6-c-menu-toggle pf-m-split-button"
      >
        <label
          class="pf-v6-c-check pf-m-standalone"
        >
          <input
            aria-invalid="false"
            aria-label="Select page"
            class="pf-v6-c-check__input"
            data-ouia-component-id="BulkSelect-checkbox"
            data-ouia-component-type="PF6/MenuToggleCheckbox"
            data-ouia-safe="true"
            name="BulkSelect-checkbox"
            type="checkbox"
          />
        </label>
        <span
          data-ouia-component-id="BulkSelect-text"
        >
          2 selected
        </span>
        <button
          aria-expanded="false"
          aria-label="Bulk select toggle"
          class="pf-v6-c-menu-toggle__button"
          data-ouia-component-id="OUIA-Generated-MenuToggle-1"
          data-ouia-component-type="PF6/MenuToggle"
          data-ouia-safe="true"
          type="button"
        >
          <span
            class="pf-v6-c-menu-toggle__controls"
          >
            <span
              class="pf-v6-c-menu-toggle__toggle-icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v6-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                />
              </svg>
            </span>
          </span>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pf-v6-c-menu-toggle pf-m-split-button"
    >
      <label
        class="pf-v6-c-check pf-m-standalone"
      >
        <input
          aria-invalid="false"
          aria-label="Select page"
          class="pf-v6-c-check__input"
          data-ouia-component-id="BulkSelect-checkbox"
          data-ouia-component-type="PF6/MenuToggleCheckbox"
          data-ouia-safe="true"
          name="BulkSelect-checkbox"
          type="checkbox"
        />
      </label>
      <span
        data-ouia-component-id="BulkSelect-text"
      >
        2 selected
      </span>
      <button
        aria-expanded="false"
        aria-label="Bulk select toggle"
        class="pf-v6-c-menu-toggle__button"
        data-ouia-component-id="OUIA-Generated-MenuToggle-1"
        data-ouia-component-type="PF6/MenuToggle"
        data-ouia-safe="true"
        type="button"
      >
        <span
          class="pf-v6-c-menu-toggle__controls"
        >
          <span
            class="pf-v6-c-menu-toggle__toggle-icon"
          >
            <svg
              aria-hidden="true"
              class="pf-v6-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
